packages/job-worker/src/blueprints/context/adlibActions.ts

```diff
@@ -64,9 +65,72 @@ export enum ActionPartChange {
 	SAFE_CHANGE = 1,
 }
 
+export class DatastoreActionExecutionContext
+	extends ShowStyleUserContext
+	implements IDataStoreActionExecutionContext, IEventContext
+{
+	protected readonly _context: JobContext
+
+	constructor(
+		contextInfo: UserContextInfo,
+		context: JobContext,
+		showStyle: ReadonlyDeep<ShowStyleCompound>,
+		showStyleBlueprintConfig: ProcessedShowStyleConfig,
+		watchedPackages: WatchedPackagesHelper
+	) {
+		super(
+			contextInfo,
+			context.studio,
+			context.getStudioBlueprintConfig(),
+			showStyle,
+			showStyleBlueprintConfig,
+			watchedPackages
+		)
+		this._context = context
+	}
+
+	async setTimelineDatastoreValue(key: string, value: unknown, mode: 'temporary' | 'indefinite'): Promise<void> {
+		const collection = this._context.directCollections.TimelineDatastores
+		const entry = await collection.findOne({ studioId: this._context.studioId, key })
+
+		if (!entry) {
+			await collection.insertOne({
+				_id: getRandomId(),
+				studioId: this._context.studioId,
+
+				key,
+				value,
+
+				modified: Date.now(),
+				mode,
+			})
+		} else {
+			await collection.update(entry._id, {
+				$set: {
+					value,
+					mode,
+					modified: Date.now(),
+				},
+			})
+		}
+	}
+
+	async removeTimelineDatastoreValue(key: string): Promise<void> {
+		const collection = this._context.directCollections.TimelineDatastores
+		await collection.remove({ studioId: this._context.studioId, key })
```

Comment:
ref: my comment on line 94:
instead do a lookup using the hash-id, otherwise this would cause a full-db-search.